### ISSUE TITLE
C# bool array does not seem to work correctly

### ISSUE THREAD

**Issue by gpetrou** [swig/swig#616]
Doing some tests with int and bool arrays it seems that int arrays work well, but bool do not.
The test method does not seem to copy the result correctly.

C#
`

```
     int[] intSource = { 1, 2, 3 };
     int[] intTarget = new int[intSource.Length];

     bool[] boolSource = { true, true, false, true, true, false, true };
     bool[] boolTarget = new bool[boolSource.Length];

     TestClass.myArrayCopyInt(intSource, intTarget, intTarget.Length);
     TestClass.myArrayCopyBool(boolSource, boolTarget, boolTarget.Length);

      foreach (int i in intTarget)
             Console.Write("{0} ", i);

      foreach (bool i in boolTarget)
            Console.Write("{0} ", i);
```

`

c++
`

```
   class TESTClass
    {
    static void myArrayCopy( int* intSource, int* intTarget, int nitems );
    static void myArrayCopy1( bool* boolSource, bool* boolTarget, int nitems );
  }

  void TestClass::myArrayCopy( int* intSource, int* intTarget, int nitems  ) {
      for (int i = 0; i < nitems; i++ ) {
     intTarget[ i ] = intSource[ i ];
   }
  }

  void TestClass::myArrayCopy1( bool* boolSource, bool* boolTarget, int nitems  ){
          for (int i = 0; i < nitems; i++ ) {
            boolTarget[ i ] = boolSource i ];
          }
    }
```

`

interface
`
         %module TestClassWrapper

```
     %{
         TestClass.h"
      %}  

       %include <wchar.i>
       %include <typemaps.i>
        %include <windows.i>
          %include <std_vector.i>
         %include "arrays_csharp.i"

      %apply int INPUT[]  { int* intSource }
       %apply int OUTPUT[] { int* intTarget }

    %apply bool INPUT[]  { bool* boolSource }
    %apply bool OUTPUT[] { bool* boolTarget }

      %include "TestClass.h"  
```

`

**Comment by myd7349** [comment]
TestClass.h
```cpp
class TestClass
{
public:
    static void myArrayCopy(int* intSource, int* intTarget, int nitems);
    static void myArrayCopy1(bool* boolSource, bool* boolTarget, int nitems);
};
```

**Comment by myd7349** [comment]
SWIG's boolean array typemaps is added after issue #229 . It is problematic, however.

**Comment by geefr** [comment]
I bumped into this one recently, and discovered that using ArraySubType doesn't work on mono (Version 'Debian 4.6.2.7'). Looks like the marshalling code ignores the subtype and just uses the default 4-byte version.

Works correctly on Microsofts implementation of .Net 4.x as far as I can tell.

If you're wanting a finished test case/fix I'm happy to pull it together, leaving the default as it is doesn't seem sensible to me as it's unlikely to be the desired behaviour for any case?

If there's a way to detect if running/built for mono there could at least be a warning/error here to stop others falling into the trap. Would simply checking for MSVC in C++ work for this, or is it possible to use gcc with the microsoft runtime?

**Comment by wsfulton** [comment]
The Microsoft implementation would be the best reference and I suggest SWIG is modified to work with this implementation. How about raising a bug report for Mono to support ArraySubType?

**Comment by geefr** [comment]
Updates/test case in #1576

**Comment by wsfulton** [comment]
Fixed, see #1576.
